File: problems/Python/30. Substring with Concatenation of All Words.py
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        if not s or not words:
            return []
            
        wlen = len(words[0])
        n = len(words)
        slen = len(s)
        tlen = wlen * n
        res = []
        wmap = {}
        for w in words:
            wmap[w] = wmap.get(w, 0) + 1

        for i in range(wlen):
            l = i
            smap = {}
            cnt = 0
            for r in range(i, slen - wlen + 1, wlen):
                word = s[r:r+wlen]
                if word in wmap:
                    smap[word] = smap.get(word, 0) + 1
                    cnt += 1
                    while smap[word] > wmap[word]:
                        lword = s[l:l+wlen]
                        smap[lword] -= 1
                        cnt -= 1
                        l += wlen
                    if cnt == n:
                        res.append(l)
                else:
                    smap.clear()
                    cnt = 0
                    l = r + wlen
        return res

File: problems/C++/30. Substring with Concatenation of All Words.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        if (s.empty() || words.empty()) return {};

        int wlen = words[0].length();
        int n = words.size();
        int slen = s.length();
        int tlen = wlen * n;
        vector<int> res;
        unordered_map<string, int> wmap;

        for (const string& w : words) {
            wmap[w]++;
        }

        for (int i = 0; i < wlen; ++i) {
            unordered_map<string, int> smap;
            int l = i;
            int cnt = 0;
            for (int r = i; r <= slen - wlen; r += wlen) {
                string word = s.substr(r, wlen);
                if (wmap.count(word)) {
                    smap[word]++;
                    cnt++;
                    while (smap[word] > wmap[word]) {
                        string lword = s.substr(l, wlen);
                        smap[lword]--;
                        cnt--;
                        l += wlen;
                    }
                    if (cnt == n) {
                        res.push_back(l);
                    }
                } else {
                    smap.clear();
                    cnt = 0;
                    l = r + wlen;
                }
            }
        }
        return res;
    }
};